File: 프로그래머스/Python/l2_더 맵게.py
import heapq

def solution(scoville, k):

    heapq.heapify(scoville)
    cnt = 0

    while len(scoville)>1:
        first = heapq.heappop(scoville)

        if first >= k:
            heapq.heappush(scoville, first)
            break
        
        second = heapq.heappop(scoville)

        new = first + second*2
        heapq.heappush(scoville, new)
        cnt+=1
    
    check = heapq.heappop(scoville)
    if check < k:
        return -1
    return cnt


solution([1, 2, 3, 9, 10, 12], 7)











# import heapq

# def solution(scoville, k):
#     heapq.heapify(scoville)
#     cnt = 0
#     while len(scoville) >1:
#         first = heapq.heappop(scoville)
#         if first >= k:
#             heapq.heappush(scoville, first)
#             heapq.heappush
#             break
#         second = heapq.heappop(scoville)
#         heapq.heappush(scoville, first + 2*second)
#         cnt +=1
    
#     check = heapq.heappop(scoville)
#     if check < k:
#         return -1
#     return cnt

File: 프로그래머스/JAVA/l2_더맵게.java
import java.util.PriorityQueue;

public class l2_더맵게 {
    public static void main(String[] args){
        int[] scoville = {1,2,3,9,10,12};
        int K = 7;
        int answer = solution(scoville, K);
    }


    public static int solution(int[] scoville, int K){
        int count = 0;
        PriorityQueue<Integer> queue = new PriorityQueue<>();
        for (int food : scoville){
            queue.add(food);
        }

        while (queue.peek() <K){
            if (queue.size() ==1)
                return -1;
            count ++;
            int first = queue.poll();
            int second = queue.poll();

            int newFood = first + second*2;
            queue.add(newFood);
            // System.out.println(count+" "+newFood);
        }


        return count;
    }
}
